src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs

```diff
@@ -127,81 +135,85 @@ public class ImageSharpMiddleware
+                IImageResolver resolver = this.resolvers.FirstOrDefault(r => r.Match(context));
 
-            // Not cached? Let's get it from the image resolver.
-            byte[] inBuffer = null;
-            byte[] outBuffer = null;
-            MemoryStream outStream = null;
-            try
-            {
-                inBuffer = await resolver.ResolveImageAsync(context, this.logger);
-                if (inBuffer == null || inBuffer.Length == 0)
+                if (resolver == null || !await resolver.IsValidRequestAsync(context, this.logger))
                 {
-                    // Log the error but let the pipeline handle the 404
-                    this.logger.LogImageResolveFailed(imageContext.GetDisplayUrl());
+                    // Nothing to do. call the next delegate/middleware in the pipeline
                     await this.next(context);
                     return;
                 }
 
-                // No allocations here for inStream since we are passing the buffer.
-                // TODO: How to prevent the allocation in outStream? Passing a pooled buffer won't let stream grow if needed.
-                outStream = new MemoryStream();
-                using (var image = FormattedImage.Load(this.options.Configuration, inBuffer))
+                CachedInfo info = await this.cache.IsExpiredAsync(key, DateTime.UtcNow.AddDays(-this.options.MaxCacheDays));
+
+                var imageContext = new ImageContext(context, this.options);
+
+                if (!info.Expired)
                 {
-                    image.Process(this.logger, this.processors, commands);
-                    this.options.OnBeforeSave?.Invoke(image);
-                    image.Save(outStream);
+                    // Image is a cached image. Return the correct response now.
+                    await this.SendResponse(imageContext, key, info.LastModifiedUtc, null, (int)info.Length);
+                    return;
                 }
 
-                // Allow for any further optimization of the image. Always reset the position just in case.
-                outStream.Position = 0;
-                this.options.OnProcessed?.Invoke(new ImageProcessingContext(context, outStream, Path.GetExtension(key)));
-                outStream.Position = 0;
-                int outLength = (int)outStream.Length;
+                // Not cached? Let's get it from the image resolver.
+                byte[] inBuffer = null;
+                byte[] outBuffer = null;
+                MemoryStream outStream = null;
+                try
+                {
+                    inBuffer = await resolver.ResolveImageAsync(context, this.logger);
+                    if (inBuffer == null || inBuffer.Length == 0)
+                    {
+                        // Log the error but let the pipeline handle the 404
+                        this.logger.LogImageResolveFailed(imageContext.GetDisplayUrl());
+                        await this.next(context);
```

Comment:
as above should be moved outside the lock... might be easier to move this whole method down into a new internal method that returns true if request was handled else false... then we can call `this.next(context)` based on the outcome instead of inline inside the lock like we have here.

Reply:
Done 👍